sdk/pinocchio/src/sysvars/slot_hashes.rs

```diff
@@ -0,0 +1,420 @@
+    ///
+    /// Checks that the buffer length is 8 + (N * 40) for some N ≤ 512.
+    #[inline]
+    fn validate_buffer_size(buffer_len: usize) -> Result<(), ProgramError> {
+        validate_slothashes_constraints(buffer_len, None)?;
+        Ok(())
+    }
+
+    /// Creates a `SlotHashes` instance from arbitrary data with full validation.
+    ///
+    /// This constructor performs comprehensive validation of the data format
+    /// including length prefix, entry count bounds, and buffer size requirements.
+    /// Does not validate that entries are sorted in descending order.
+    pub fn new(data: T) -> Result<Self, ProgramError> {
+        let num_entries = parse_and_validate_data(&data)?;
+        Ok(unsafe { Self::new_unchecked(data, num_entries) })
+    }
+
+    /// Creates a `SlotHashes` instance directly from a data container and entry count.
+    /// Important: provide a valid len. Whether or not len is assumed to be
+    /// the constant 20_488 (512 entries) is up to caller.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because it does not check the validity of the data or count.
+    /// The caller must ensure:
+    /// 1. The underlying byte slice in `data` represents valid SlotHashes data
+    ///    (length prefix + entries, where entries are sorted in descending order by slot).
+    /// 2. `len` is the correct number of entries (≤ MAX_ENTRIES), matching the prefix.
+    /// 3. The data slice contains at least `NUM_ENTRIES_SIZE + len * ENTRY_SIZE` bytes.
+    /// 4. Alignment is correct for SlotHashEntry access.
+    ///
+    #[inline]
+    pub unsafe fn new_unchecked(data: T, len: usize) -> Self {
+        debug_assert!(len <= MAX_ENTRIES && data.len() >= NUM_ENTRIES_SIZE + len * ENTRY_SIZE);
+        SlotHashes { data, len }
+    }
+
+    /// Gets the number of entries stored in this SlotHashes instance.
+    /// Performs validation checks and returns the entry count if valid.
+    pub fn get_entry_count(&self) -> Result<usize, ProgramError> {
+        let data_entry_count = read_entry_count_from_bytes(&self.data).unwrap_or(0);
+        if data_entry_count != self.len {
+            return Err(ProgramError::InvalidArgument);
+        }
+        Ok(self.len)
+    }
+
+    /// Reads the entry count directly from the beginning of this SlotHashes instance **without validation**.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because it performs no checks on the underlying data.
+    /// The caller **must** ensure that:
+    /// 1. The underlying data contains at least `NUM_ENTRIES_SIZE` (8) bytes.
+    /// 2. The first 8 bytes represent a valid `u64` in little-endian format.
+    /// 3. Calling this function without ensuring the above may lead to panics
+    ///    (out-of-bounds access) or incorrect results.
+    #[inline(always)]
+    pub unsafe fn get_entry_count_unchecked(&self) -> usize {
+        read_entry_count_from_bytes_unchecked(&self.data)
```

Comment:
Should this be `self.len`?

Reply:
Hmm, interesting. `new_unchecked` makes no promise at all that `len` of the data matches the header of the sysvar data - so here, get_entry_count_unchecked() represents getting what the data really says rather than what the caller assumed about it when constructing.

Since 512 entries is golden on mainnet and we're not really in the business of questioning upstream invariants, I'm fine with `len`.

But if `get_entry_count_unchecked` is just a wrapper for `len`, it should just be removed. And maybe we could imagine other scenarios where it's useful to just read the header.